### ISSUE TITLE
phf_map! (sort of) on stable with macro_rules + proc_macro_derive

### ISSUE THREAD

**Issue by SimonSapin** [sfackler/rust-phf#106]
You may be interested in this: https://github.com/servo/rust-cssparser/blob/48905424c5214b22a545eb7135cf671fbd63c312/src/lib.rs#L158-L211

I’ve managed to make `macro_rules!` macro similar to `phf_map!`, that uses a derive procedural macro internally instead of a compiler plugin,and therefore runs on Rust 1.15 stable.

Differences are:

* The macro expands to an item, not an expression. (In the `phf_map` plugin, it expands to a const expression used to initialize a `static`.) Extra work is needed to support `pub` v.s. not, doc-comments, etc.
* Values are given as strings that contain Rust syntax for const expressions, rather than literally as const expressions.

Both are limitations due to busing `derive` rather than having a proper functional procedural macro (which are not stable yet).

(This particular macro also does case-insensitive matching, but that part can easily be removed.)

**Comment by cbreeden** [comment]
It's worth noting that @SimonSapin was able to remove the Values are given as strings using the proc-macro-hack <https://github.com/servo/rust-cssparser/blob/master/src/macros.rs#L55-L100>.

**Comment by SimonSapin** [comment]
Permanent link: https://github.com/servo/rust-cssparser/blob/1b1f7033d387608470a8d851a5b9076d557451b8/src/macros.rs#L55-L100

This is enabled by https://docs.rs/procedural-masquerade/

**Comment by cbreeden** [comment]
Another note: you can rewrite the example so that `phf_map!` expands to an expression as explained here: <https://docs.rs/proc-macro-hack/0.3.3/proc_macro_hack/#expansion-of-expression-macros>.  The idea that the procedural macro defines a macro named `proc_macro_call!()` which is then invoked during expansion.

**Comment by abonander** [comment]
Now that the other proc macros are being stabilized, we don't need to consider these workarounds. It should be relatively straightforward now to port `phf_map! {}` to a `#[proc_macro]`.

**Comment by sfackler** [comment]
Expression macros aren't being stabilized, but we could have a lazy_static! style API like

```rust
phf_map! {
    pub static MAP: PhfMap<&'static str, u32> = {
        "1": 1,
        "2": 2,
    };
}
```

**Comment by abonander** [comment]
The above mentioned workarounds should still work then, a trivial wrapper that invokes `phf_map!` to produce a `const` which is then returned as an rvalue. I'm not sure if it's possible to have a `macro_rules` macro invoke a proc macro without requiring the latter to be imported in the same scope, however. It's worth a try.

**Comment by kurnevsky** [comment]
Expression macros can be defined on stable rust now via https://github.com/dtolnay/proc-macro-hack

**Comment by abonander** [comment]
Proc macros are stable now, I think someone just needs to implement it.

**Comment by abonander** [comment]
Closed by the merge of #149, will be in 0.8.0.
